### PR TITLE
ENH: Add IFU compatibility for SPHERE

### PR DESCRIPTION
**Pull request by DrSoulain**
As noted in #145, the cleaning step of AMICAL is not compatible with the standard IFU data obtained with VLT/SPHERE. For now, we add some compatibility keywords to deal with such types of data. The main idea is to process each spectral channel individually by the use of `i_wl` keyword.

### CODE REVIEW

File: amical/mf_pipeline/ami_function.py [hunk 1]
```diff
@@ -385,6 +385,12 @@ def make_mf(
     filt = get_wavelength(instrument, filtname)
 
     if instrument == "SPHERE-IFS":
+        if i_wl is None:
+            raise ValueError(
+                "Your file seems to be obtained with IFU instrument: spectral "
+                + f"channel index `i_wl` must be specified (nlambda = {len(filt)}) and "
+                + "should be strictly identical to the one used for the cleaning step."
```
Comment (neutrinoceros) [inline amical/mf_pipeline/ami_function.py:392]: `+` is not necessary here

File: amical/data_processing.py [hunk 2]
```diff
@@ -279,11 +279,11 @@ def sky_correction(imA, r1=None, dr=None, verbose=False, *, center=None, mask=No
...
-    elif verbose:
-        print(
-            f"Sky correction of {backgroundB} was subtracted,"
-            f" remaining background is {backgroundC}."
-        )
+    # elif verbose:
+    #     print(
+    #         f"Sky correction of {backgroundB} was subtracted,"
+    #         f" remaining background is {backgroundC}."
+    #     )
```
Comment (neutrinoceros) [inline amical/data_processing.py:286]:
```suggestion
```

File: amical/data_processing.py [hunk 3]
```diff
@@ -376,6 +374,7 @@ def show_clean_params(
     offy=0,
     apod=False,
     window=None,
+    ifu=False,
     *,
```
Comment (neutrinoceros) [inline amical/data_processing.py:378]: New arguments should always go after `*,`
```suggestion
    *,
    ifu=False,
```

File: amical/data_processing.py [hunk 4]
```diff
@@ -725,6 +730,20 @@ def select_clean_data(
...
 
+        n_axis = len(cube.shape)
+        if n_axis == 4:
+            ifu = True
+            naxis4 = hdr["NAXIS4"]
+            if i_wl is None:
+                raise ValueError(
+                    "Your file seems to be obtained with IFU data: spectral "
+                    + f"channel index `i_wl` must be specified (nlambda = {naxis4})."
+                    + ""
```
Comment (neutrinoceros) [inline amical/data_processing.py:741]:
```suggestion
                    f"channel index `i_wl` must be specified (nlambda = {naxis4})."
```

File: amical/get_infos_obs.py [hunk 5]
```diff
@@ -231,3 +231,79 @@ def get_pixel_size(ins):
...
+):
+    """Get spectral information for the given instrumental IFU setup.
+    `i_wl` can be an integer, a list of 2 integers (to get a range between those
+    two) or a list of integers (>= 3) used to display the
+    requested spectral channels."""
+    from matplotlib import pyplot as plt
+
+    wl = get_wavelength(instrument, filtname) * 1e6
+
+    if np.isnan(wl.any()):
```
Comment (neutrinoceros) [inline amical/get_infos_obs.py:247]: what ? doesn't `wl.any()` returns a boolean ? in what scenario could its return value be `np.nan` ?

File: amical/get_infos_obs.py [hunk 6]
```diff
@@ -231,3 +231,79 @@ def get_pixel_size(ins):
...
+
+
+def get_ifu_table(
+    i_wl, filtname="YH", instrument="SPHERE-IFS", verbose=False, display=False
+):
+    """Get spectral information for the given instrumental IFU setup.
+    `i_wl` can be an integer, a list of 2 integers (to get a range between those
+    two) or a list of integers (>= 3) used to display the
+    requested spectral channels."""
+    from matplotlib import pyplot as plt
```
Comment (neutrinoceros) [inline amical/get_infos_obs.py:243]: this import isn't used unless `display == True`, it should be moved accordingly

File: amical/get_infos_obs.py [hunk 7]
```diff
@@ -231,3 +231,79 @@ def get_pixel_size(ins):
...
+        print(f"\nInstrument: {instrument}, spectral range: {filtname}")
+        print("-----------------------------")
+        print(
+            "spectral coverage: %2.2f - %2.2f µm (step = %2.2f)"
+            % (wl[0], wl[-1], np.diff(wl)[0])
+        )
+
+    one_wl = True
+    multiple_wl = False
+    if isinstance(i_wl, list) & (len(i_wl) == 2):
```
Comment (neutrinoceros) [inline amical/get_infos_obs.py:260]: `and` isn't syntactic sugar for `&`, they are fundamentally different operators.
```suggestion
    if isinstance(i_wl, list) and len(i_wl) == 2:
```

File: amical/get_infos_obs.py [hunk 8]
```diff
@@ -231,3 +231,79 @@ def get_pixel_size(ins):
...
+        one_wl = False
+        wl_range = wl[i_wl[0] : i_wl[1]]
+        sp_range = np.arange(i_wl[0], i_wl[1], 1)
+    elif isinstance(i_wl, list) & (len(i_wl) > 2):
```
Comment (neutrinoceros) [inline amical/get_infos_obs.py:264]:
```suggestion
    elif isinstance(i_wl, list) and len(i_wl) > 2:
```

File: amical/mf_pipeline/ami_function.py (hunk 1 above)
Comment (neutrinoceros) [inline amical/mf_pipeline/ami_function.py:392]:
```suggestion
                f"channel index `i_wl` must be specified (nlambda = {len(filt)}) and "
                "should be strictly identical to the one used for the cleaning step."
```

Comment (neutrinoceros) [reply amical/mf_pipeline/ami_function.py:392]: More importantly, I don't understand this message. It's raised on a condition that depends only on parameters the function receives directly, but it's pointing at a problem somewhere else ?

File: amical/oifits.py [hunk 9]
```diff
@@ -1013,6 +1013,146 @@ def _plot_CP(ax3, l_dic, dic_color, conv_cp, diffWl=False):
...
+        U = tmp.U
+        V = tmp.V
+        wl = tmp.wl
+        l_bmax.append(tmp.bmax)
+        all_U.append(U)
+        all_V.append(V)
+        all_wl.append([wl] * len(U))
+    all_U = np.array(all_U)
+    all_V = np.array(all_V)
+    all_wl = np.array(all_wl) * 1e6
```
Comment (neutrinoceros) [inline amical/oifits.py:1031]: this is a very inefficient way to construct arrays with deterministic sizes. I know performance isn't a common concern in AMICAL but this stands out to me: we can make it much quicker.

File: amical/oifits.py [hunk 10]
```diff
@@ -1013,6 +1013,146 @@ def _plot_CP(ax3, l_dic, dic_color, conv_cp, diffWl=False):
...
+    ax1.scatter(
+        all_U,
+        all_V,
+        s=40,
+        c=all_wl,
+        marker="o",
+        edgecolors="#364f6b",
+        alpha=1,
+        linewidth=0.1,
+        cmap="jet",
```
Comment (neutrinoceros) [inline amical/oifits.py:1042]: I mean, really ?
```suggestion
        cmap="turbo",
```

File: amical/oifits.py [hunk 11]
```diff
@@ -1013,6 +1013,146 @@ def _plot_CP(ax3, l_dic, dic_color, conv_cp, diffWl=False):
...
+    sc = ax1.scatter(
+        -all_U,
+        -all_V,
+        s=40,
+        c=all_wl,
+        marker="o",
+        edgecolors="#364f6b",
+        alpha=1,
+        linewidth=0.1,
+        cmap="jet",
```
Comment (neutrinoceros) [inline amical/oifits.py:1053]:
```suggestion
        cmap="turbo",
```

File: amical/oifits.py [hunk 12]
```diff
@@ -1013,6 +1013,146 @@ def _plot_CP(ax3, l_dic, dic_color, conv_cp, diffWl=False):
...
+        all_freq,
+        all_V2,
+        s=20,
+        c=all_wl,
+        zorder=20,
+        linewidth=0.5,
+        marker="o",
+        edgecolors="#364f6b",
+        alpha=1,
+        cmap="jet",
```
Comment (neutrinoceros) [inline amical/oifits.py:1107]:
```suggestion
        cmap="turbo",
```

File: amical/oifits.py [hunk 13]
```diff
@@ -1013,6 +1013,146 @@ def _plot_CP(ax3, l_dic, dic_color, conv_cp, diffWl=False):
...
+        tmp = _apply_flag(dic, unit="arcsec")
+        cp = tmp.cp * conv_cp
+        e_cp = tmp.e_cp * conv_cp
+        sp_freq_cp = tmp.sp_freq_cp
+        max_f_cp.append(np.max(sp_freq_cp))
+
+        all_CP.extend(cp)
+        all_e_cp.extend(e_cp)
+        all_freq.extend(sp_freq_cp)
+        all_wl.extend([tmp.wl] * len(cp))
```
Comment (neutrinoceros) [inline amical/oifits.py:1126]: see my previous comment about performance here

File: amical/oifits.py [hunk 14]
```diff
@@ -1013,6 +1013,146 @@ def _plot_CP(ax3, l_dic, dic_color, conv_cp, diffWl=False):
...
+    all_V2 = np.array(all_V2)
+    all_e_V2 = np.array(all_e_V2)
+    all_freq = np.array(all_freq)
+    all_wl = np.array(all_wl)
+
+    ax2.errorbar(
+        all_freq,
+        all_V2,
+        yerr=all_e_V2,
+        linestyle="None",
```
Comment (neutrinoceros) [inline amical/oifits.py:1088]: matplotlib has a history of being too flexible with what counts as a None, or what counts as a string, but let's not abuse it
```suggestion
        linestyle=None,
```

File: amical/oifits.py [hunk 15]
```diff
@@ -1013,6 +1013,146 @@ def _plot_CP(ax3, l_dic, dic_color, conv_cp, diffWl=False):
...
+        capsize=1,
+        ecolor="#364f6b",
+        mec="#364f6b",
+        marker="None",
```
Comment (neutrinoceros) [inline amical/oifits.py:1092]:
```suggestion
        marker=None,
```

File: amical/oifits.py [hunk 16]
```diff
@@ -1013,6 +1013,146 @@ def _plot_CP(ax3, l_dic, dic_color, conv_cp, diffWl=False):
...
+        all_freq,
+        all_CP,
+        s=20,
+        c=all_wl,
+        zorder=20,
+        linewidth=0.5,
+        marker="o",
+        edgecolors="#364f6b",
+        alpha=1,
+        cmap="jet",
```
Comment (neutrinoceros) [inline amical/oifits.py:1151]:
```suggestion
        cmap="turbo",
```

File: amical/tests/test_io.py [hunk 17]
```diff
@@ -314,3 +339,82 @@ def test_quiet_mode(global_datadir, capsys):
...
+        targetname="test",
+        bs_multi_tri=False,
+        maskname="g7",
+        filtname="YH",
+        fw_splodge=0.7,
+        display=False,
+    )
+    bs_keys = list(bs.keys())
+    assert isinstance(bs, munch.Munch)
+    assert len(bs_keys) == 13
```
Comment (neutrinoceros) [inline amical/tests/test_io.py:363]:
```suggestion
    assert isinstance(bs, munch.Munch)
    assert len(bs) == 13
```

File: amical/tests/test_io.py [hunk 18]
```diff
@@ -314,3 +339,82 @@ def test_quiet_mode(global_datadir, capsys):
...
+            cube_dirty,
+            fits_file,
+            targetname="test",
+            bs_multi_tri=False,
+            maskname="g7",
+            filtname="YH",
+            fw_splodge=0.7,
+            display=False,
+        )
+    assert exc_info.value.args[0] == msg_error
```
Comment (neutrinoceros) [inline amical/tests/test_io.py:389]:
```suggestion
    with pytest.raises(
        ValueError,
        match= (
            "Your file seems to be obtained with IFU instrument: spectral channel "
            f"index `i_wl` must be specified (nlambda = {nlambda}) and should be strictly"
            " identical to the one used for the cleaning step."
        )
    ):
        amical.extract_bs(
            cube_dirty,
            fits_file,
            targetname="test",
            bs_multi_tri=False,
            maskname="g7",
            filtname="YH",
            fw_splodge=0.7,
            display=False,
        )
```

File: amical/tests/test_tools.py [hunk 19]
```diff
@@ -104,3 +105,25 @@ def test_compute_pa_niriss_amisim(global_datadir):
         pa = tools.compute_pa(hdr, n_ps)
     assert len(record) == 1
     assert pa == 0.0
+
+
+@pytest.mark.usefixtures("close_figures")
+@pytest.mark.parametrize("list_index_ifu", [[0], [0, 10], [0, 1, 2]])
+@pytest.mark.parametrize("filtname", ["YJ", "YH"])
+def test_get_table_ifu(list_index_ifu, filtname):
+    from matplotlib import pyplot as plt
```
Comment (neutrinoceros) [inline amical/tests/test_tools.py:114]: nested imports are useful in the application code to reduce startup time, but they don't have much value in test code

File: amical/data_processing.py (hunk 2 above)
Comment (DrSoulain) [reply amical/data_processing.py:286]: I just remove this comment to have a clean progress bar if used with verbose. But actually, how can you deal with tqdm when you have such print between iteration?

File: amical/get_infos_obs.py (hunk 5 above)
Comment (DrSoulain) [reply amical/get_infos_obs.py:247]: Indeed, now we deal the potential issue within get_wavelength(), so no need of this check anymore.

File: amical/mf_pipeline/ami_function.py (hunk 1 above)
Comment (DrSoulain) [reply amical/mf_pipeline/ami_function.py:392]: Yes, you need to use the same `i_wl` argument in both function (clean and extract). That's why I added this message but it's maybe useless to detail this point there.

File: amical/data_processing.py (hunk 2 above)
Comment (neutrinoceros) [reply amical/data_processing.py:286]: I don't know if that's possible. That's probably something you could do with `rich.progress` but that would be whole new dependency and migration would be a much bigger undertaking

File: amical/get_infos_obs.py [hunk 20]
```diff
@@ -231,3 +231,79 @@ def get_pixel_size(ins):
...
+def get_ifu_table(
+    i_wl, filtname="YH", instrument="SPHERE-IFS", verbose=False, display=False
+):
+    """Get spectral information for the given instrumental IFU setup.
+    `i_wl` can be an integer, a list of 2 integers (to get a range between those
+    two) or a list of integers (>= 3) used to display the
+    requested spectral channels."""
+    wl = get_wavelength(instrument, filtname) * 1e6
+
+    if np.isnan(wl.any()):
```
Comment (neutrinoceros) [inline amical/get_infos_obs.py:245]:
```suggestion
    if np.any(np.isnan(a))
```

File: amical/oifits.py [hunk 21]
```diff
@@ -1112,14 +1252,11 @@ def show(
...
-    # Work in progress
-    # ins = l_dic[0]["info"]["INSTRUME"]
-    # if ("IFS" in ins) & (len(l_dic) > 1):
-    #     l_bmax = _plot_UV_ifu(ax1, fig, l_dic)
-    # else:
-    # Work in progress
-
-    l_bmax = _plot_UV(ax1, l_dic, dic_color, diffWl=False)
+    ins = l_dic[0]["info"]["INSTRUME"]
+    if ("IFS" in ins) & (len(l_dic) > 1):
```
Comment (neutrinoceros) [inline amical/oifits.py:1256]:
```suggestion
    if "IFS" in ins and len(l_dic) > 1:
```

File: amical/oifits.py [hunk 22]
```diff
@@ -1159,14 +1296,10 @@ def show(
...
     # -------
-
-    # Work in progress
-    # if ("IFS" in ins) & (len(l_dic) > 1):
-    #     max_f_vis = _plot_V2_ifu(ax2, l_dic)
-    # else:
-    # Work in progress
-
-    max_f_vis = _plot_V2(ax2, l_dic, dic_color, diffWl=diffWl)
+    if ("IFS" in ins) & (len(l_dic) > 1):
```
Comment (neutrinoceros) [inline amical/oifits.py:1299]:
```suggestion
    if "IFS" in ins and len(l_dic) > 1:
```

File: amical/oifits.py [hunk 23]
```diff
@@ -1197,13 +1330,10 @@ def show(
...
     cmin = -cmax
 
-    # Work in progress
-    # if ("IFS" in ins) & (len(l_dic) > 1):
-    #     max_f_cp = _plot_CP_ifu(ax3, l_dic, conv_cp)
-    # else:
-    # Work in progress
-
-    max_f_cp = _plot_CP(ax3, l_dic, dic_color, conv_cp, diffWl=diffWl)
+    if ("IFS" in ins) & (len(l_dic) > 1):
```
Comment (neutrinoceros) [inline amical/oifits.py:1333]:
```suggestion
    if "IFS" in ins and len(l_dic) > 1:
```

File: amical/tests/test_io.py [hunk 24]
```diff
@@ -37,6 +39,31 @@ def test_load_file(example_oifits):
...
+@pytest.fixture()
+def ifs_ami_param():
+    params_ami = {
+        "peakmethod": "fft",
+        "bs_multi_tri": False,
+        "maskname": "g7",
+        "fw_splodge": 0.7,
+        "filtname": "YJ",
+    }
+    return params_ami
```
Comment (neutrinoceros) [inline amical/tests/test_io.py:64]:
```suggestion
    return {
        "peakmethod": "fft",
        "bs_multi_tri": False,
        "maskname": "g7",
        "fw_splodge": 0.7,
        "filtname": "YJ",
    }
```

**Review by neutrinoceros — APPROVED** [review]

**Pull request merged**